This patch makes it possible to build cmd/dot without haing Pango on a system.

Additionally, it cleans up the massive duplication that was taking place in
cmd/dot/Makefile.am, without affecting the end result.

Tested with:
- external libGD + no system cairo/pango
- external libGD + system cairo/pango

I did not test internal combinations using internal libGD, but I'm reasonably
certain they will work fine.

Signed-off-by: Robin H. Johnson <robbat2@gentoo.org>
Date: 20 December 2007

diff -Nuar -NuarwbB graphviz-2.16.1.orig/cmd/dot/Makefile.am graphviz-2.16.1/cmd/dot/Makefile.am
--- graphviz-2.16.1.orig/cmd/dot/Makefile.am	2007-12-20 18:23:03.354935000 -0800
+++ graphviz-2.16.1/cmd/dot/Makefile.am	2007-12-20 18:48:09.566483255 -0800
@@ -67,91 +67,84 @@
 # Workaround is to link explicity to the .libs/libxxx.a
 # A better approach is probably to provide an explicit list of builtins
 # like $(top_srcdir)/lib/gvc/dot_builtins.c
+
+# Static build
 dot_static_SOURCES = dot.c $(top_srcdir)/lib/gvc/no_demand_loading.c
 dot_static_LDFLAGS = -export-dynamic -dlpreopen self -export-symbols-regex '.*_LTX_library'
-if WITH_MYLIBGD
-if WITH_PANGOCAIRO
-dot_static_LDADD = \
+dot_static_LDADD_plugins = \
 	$(top_builddir)/plugin/dot_layout/.libs/libgvplugin_dot_layout_C.a \
 	$(top_builddir)/plugin/neato_layout/.libs/libgvplugin_neato_layout_C.a \
 	$(top_builddir)/plugin/core/.libs/libgvplugin_core_C.a \
-	$(top_builddir)/plugin/gd/.libs/libgvplugin_gd_C.a \
-	$(top_builddir)/plugin/pango/.libs/libgvplugin_pango_C.a \
+	$(top_builddir)/plugin/gd/.libs/libgvplugin_gd_C.a
+if WITH_PANGOCAIRO
+dot_static_LDADD_plugins += \
+	$(top_builddir)/plugin/pango/.libs/libgvplugin_pango_C.a
+endif
+
+dot_static_LDADD_lib_int = \
 	$(top_builddir)/lib/gvc/libgvc_C.la \
 	$(top_builddir)/lib/pathplan/libpathplan_C.la \
 	$(top_builddir)/lib/graph/libgraph_C.la \
-	$(top_builddir)/lib/cdt/libcdt_C.la \
+	$(top_builddir)/lib/cdt/libcdt_C.la
+dot_static_LDADD_lib_ext =
+
+if WITH_MYLIBGD
+dot_static_LDADD_lib_int += \
 	$(top_builddir)/lib/gd/libgvgd_C.la \
-       	@ICONV_LIBS@ @FONTCONFIG_LIBS@ @FREETYPE2_LIBS@ @JPEG_LIBS@ @PNG_LIBS@ @EXPAT_LIBS@ @Z_LIBS@ @LIBGEN_LIBS@ @LIBLTDL@ @SOCKET_LIBS@ @IPSEPCOLA_LIBS@ @PANGOCAIRO_LIBS@ @MATH_LIBS@
 else
-dot_static_LDADD = \
-	$(top_builddir)/plugin/dot_layout/.libs/libgvplugin_dot_layout_C.a \
-	$(top_builddir)/plugin/neato_layout/.libs/libgvplugin_neato_layout_C.a \
-	$(top_builddir)/plugin/core/.libs/libgvplugin_core_C.a \
-	$(top_builddir)/plugin/gd/.libs/libgvplugin_gd_C.a \
-	$(top_builddir)/lib/gvc/libgvc_C.la \
-	$(top_builddir)/lib/pathplan/libpathplan_C.la \
-	$(top_builddir)/lib/graph/libgraph_C.la \
-	$(top_builddir)/lib/cdt/libcdt_C.la \
-	$(top_builddir)/lib/gd/libgvgd_C.la \
-       	@ICONV_LIBS@ @FONTCONFIG_LIBS@ @FREETYPE2_LIBS@ @JPEG_LIBS@ @PNG_LIBS@ @EXPAT_LIBS@ @Z_LIBS@ @LIBGEN_LIBS@ @LIBLTDL@ @SOCKET_LIBS@ @IPSEPCOLA_LIBS@ @MATH_LIBS@
+dot_static_LDADD_lib_ext += \
+	@GD_LIBS@
 endif
-else
-dot_static_LDADD = \
-	$(top_builddir)/plugin/dot_layout/.libs/libgvplugin_dot_layout_C.a \
-	$(top_builddir)/plugin/neato_layout/.libs/libgvplugin_neato_layout_C.a \
-	$(top_builddir)/plugin/core/.libs/libgvplugin_core_C.a \
-	$(top_builddir)/plugin/gd/.libs/libgvplugin_gd_C.a \
-	$(top_builddir)/plugin/pango/.libs/libgvplugin_pango_C.a \
-	$(top_builddir)/lib/gvc/libgvc_C.la \
-	$(top_builddir)/lib/pathplan/libpathplan_C.la \
-	$(top_builddir)/lib/graph/libgraph_C.la \
-	$(top_builddir)/lib/cdt/libcdt_C.la \
-       	@GD_LIBS@ @ICONV_LIBS@ @FONTCONFIG_LIBS@ @FREETYPE2_LIBS@ @JPEG_LIBS@ @PNG_LIBS@ @EXPAT_LIBS@ @Z_LIBS@ @LIBGEN_LIBS@ @LIBLTDL@ @SOCKET_LIBS@ @IPSEPCOLA_LIBS@ @PANGOCAIRO_LIBS@ @MATH_LIBS@
+
+dot_static_LDADD_lib_ext += \
+	@ICONV_LIBS@ @FONTCONFIG_LIBS@ @FREETYPE2_LIBS@ @JPEG_LIBS@ @PNG_LIBS@ @EXPAT_LIBS@ @Z_LIBS@ @LIBGEN_LIBS@ @LIBLTDL@ @SOCKET_LIBS@ @IPSEPCOLA_LIBS@
+if WITH_PANGOCAIRO
+dot_static_LDADD_lib_ext += \
+       	@PANGOCAIRO_LIBS@
 endif
+dot_static_LDADD_lib_ext += \
+	@MATH_LIBS@
+
+dot_static_LDADD = $(dot_static_LDADD_plugins) $(dot_static_LDADD_lib_int) $(dot_static_LDADD_lib_ext)
 
+# Dynamic build
 dot_builtins_SOURCES = dot.c $(top_srcdir)/lib/gvc/dot_builtins.c $(top_srcdir)/lib/gvc/no_demand_loading.c
-if WITH_MYLIBGD
-if WITH_PANGOCAIRO
-dot_builtins_LDADD = \
+
+dot_builtins_LDADD_plugins = \
 	$(top_builddir)/plugin/dot_layout/libgvplugin_dot_layout.la \
 	$(top_builddir)/plugin/neato_layout/libgvplugin_neato_layout.la \
 	$(top_builddir)/plugin/core/libgvplugin_core.la \
-	$(top_builddir)/plugin/gd/libgvplugin_gd.la \
-	$(top_builddir)/plugin/pango/libgvplugin_pango.la \
+	$(top_builddir)/plugin/gd/libgvplugin_gd.la
+if WITH_PANGOCAIRO
+dot_builtins_LDADD_plugins += \
+	$(top_builddir)/plugin/pango/libgvplugin_pango.la
+endif
+
+dot_builtins_LDADD_lib_int = \
 	$(top_builddir)/lib/gvc/libgvc_builtins.la \
 	$(top_builddir)/lib/pathplan/libpathplan.la \
 	$(top_builddir)/lib/graph/libgraph.la \
-	$(top_builddir)/lib/cdt/libcdt.la \
-	$(top_builddir)/lib/gd/libgvgd_C.la \
-       	@ICONV_LIBS@ @FONTCONFIG_LIBS@ @FREETYPE2_LIBS@ @JPEG_LIBS@ @PNG_LIBS@ @EXPAT_LIBS@ @Z_LIBS@ @LIBGEN_LIBS@ @LIBLTDL@ @SOCKET_LIBS@ @IPSEPCOLA_LIBS@ @PANGOCAIRO_LIBS@ @MATH_LIBS@
+	$(top_builddir)/lib/cdt/libcdt.la
+dot_builtins_LDADD_lib_ext =
+
+if WITH_MYLIBGD
+dot_builtins_LDADD_lib_int += \
+	$(top_builddir)/lib/gd/libgvgd_C.la
 else
-dot_builtins_LDADD = \
-	$(top_builddir)/plugin/dot_layout/libgvplugin_dot_layout.la \
-	$(top_builddir)/plugin/neato_layout/libgvplugin_neato_layout.la \
-	$(top_builddir)/plugin/core/libgvplugin_core.la \
-	$(top_builddir)/plugin/gd/libgvplugin_gd.la \
-	$(top_builddir)/lib/gvc/libgvc_builtins.la \
-	$(top_builddir)/lib/pathplan/libpathplan.la \
-	$(top_builddir)/lib/graph/libgraph.la \
-	$(top_builddir)/lib/cdt/libcdt.la \
-	$(top_builddir)/lib/gd/libgvgd_C.la \
-       	@ICONV_LIBS@ @FONTCONFIG_LIBS@ @FREETYPE2_LIBS@ @JPEG_LIBS@ @PNG_LIBS@ @EXPAT_LIBS@ @Z_LIBS@ @LIBGEN_LIBS@ @LIBLTDL@ @SOCKET_LIBS@ @IPSEPCOLA_LIBS@ @MATH_LIBS@
+dot_builtins_LDADD_lib_ext += \
+	@GD_LIBS@
 endif
-else
-dot_builtins_LDADD = \
-	$(top_builddir)/plugin/dot_layout/libgvplugin_dot_layout.la \
-	$(top_builddir)/plugin/neato_layout/libgvplugin_neato_layout.la \
-	$(top_builddir)/plugin/core/libgvplugin_core.la \
-	$(top_builddir)/plugin/gd/libgvplugin_gd.la \
-	$(top_builddir)/plugin/pango/libgvplugin_pango.la \
-	$(top_builddir)/lib/gvc/libgvc_builtins.la \
-	$(top_builddir)/lib/pathplan/libpathplan.la \
-	$(top_builddir)/lib/graph/libgraph.la \
-	$(top_builddir)/lib/cdt/libcdt.la \
-       	@GD_LIBS@ @ICONV_LIBS@ @FONTCONFIG_LIBS@ @FREETYPE2_LIBS@ @JPEG_LIBS@ @PNG_LIBS@ @EXPAT_LIBS@ @Z_LIBS@ @LIBGEN_LIBS@ @LIBLTDL@ @SOCKET_LIBS@ @IPSEPCOLA_LIBS@ @PANGOCAIRO_LIBS@ @MATH_LIBS@
+
+dot_builtins_LDADD_lib_ext += \
+	@ICONV_LIBS@ @FONTCONFIG_LIBS@ @FREETYPE2_LIBS@ @JPEG_LIBS@ @PNG_LIBS@ @EXPAT_LIBS@ @Z_LIBS@ @LIBGEN_LIBS@ @LIBLTDL@ @SOCKET_LIBS@ @IPSEPCOLA_LIBS@
+if WITH_PANGOCAIRO
+dot_builtins_LDADD_lib_ext += \
+       	@PANGOCAIRO_LIBS@
 endif
+dot_builtins_LDADD_lib_ext += \
+	@MATH_LIBS@
 
+dot_builtins_LDADD = $(dot_builtins_LDADD_plugins) $(dot_builtins_LDADD_lib_int) $(dot_builtins_LDADD_lib_ext)
 
 EXTRA_DIST = Makefile.old dot.1 dot.pdf
 
